storage/ndb/src/kernel/blocks/dblqh/DblqhMain.cpp

```diff
@@ -21036,8 +21270,80 @@ void Dblqh::sendScanFragConf(Signal* signal,
+    FragrecordPtr regFragptr;
+    regFragptr.i = regTcPtr->fragmentptr;
+    ndbrequire(c_fragment_pool.getPtr(regFragptr));
+    if (regFragptr.p->fragId == 0 && scanPtr->m_aggregation) {
+      print = true;
+      ndbrequire(regFragptr.p->fragId == scanPtr->m_agg_interpreter->frag_id());
+      fprintf(stderr, "MOZ sendScanFragConf, rows:[%u, %u], bytes: [%u, %u], n_res_recs: %u\n",
+                                    scanPtr->m_agg_curr_batch_size_rows,
+                                    scanPtr->m_curr_batch_size_rows,
+                                    scanPtr->m_agg_curr_batch_size_bytes,
+                                    scanPtr->m_curr_batch_size_bytes,
+                                    scanPtr->m_agg_n_res_recs);
+    }
+  }
+#endif // MOZ_AGG_DEBUG
+  /*
+   * Moz
+   * In pushdown aggregation mode, before we send
+   * scanfragconf to TC, we need to make sure that
+   * there is no agg results cached in aggregation
+   * interpreter.
+   * See more details at Dblqh::scanTupkeyConfLab(),
+   * search for [MOZ-COMMENT] there.
+   */
+  if (scanPtr->m_aggregation) {
+    /*
+     * Moz
+     * In groupby mode, sendScanFragConf when:
+     *  1. batch complete: scanPtr->m_agg_interpreter->NumOfResRecords()
+     *     will at least return 1 even group by hash is empty
+     *  2. scan complete: scanPtr->m_agg_interpreter->NumOfResRecords()
+     *     will return 0;
+     * In non-group mode:
+     *  1. never reach batch complete
+     *  2. scan complete: scanPtr->m_agg_interpreter->NumOfResRecords()
+     *     will return 0;
+     *
+     * Exception:
+     * If an error happened on other fragment's aggregation interpreter,
+     * this function will be invoked by Dblqh::closeScanLab(). In this situation,
+     * since it is in a error handling process, the scanState is WAIT_CLOSE_SCAN
+     * and m_agg_n_res_recs can be bigger than 1.
+     */
+    ndbrequire(scanPtr->scanState == Dblqh::ScanRecord::WAIT_CLOSE_SCAN ||
+               (scanPtr->m_agg_n_res_recs == 0 ||
+               scanPtr->m_agg_n_res_recs == 1));
+  }
+  // Moz
+  // Make sure that we send correct m_curr_batch_size_XXX, otherwise
+  // the API cannot start to parse the TRANSID_AI message
+  Uint32 tmp_completed_ops = scanPtr->m_aggregation ?
+                                    scanPtr->m_agg_curr_batch_size_rows :
+                                    scanPtr->m_curr_batch_size_rows;
+  Uint32 tmp_total_len = scanPtr->m_aggregation ?
+                                scanPtr->m_agg_curr_batch_size_bytes :
+                                scanPtr->m_curr_batch_size_bytes;
+  ndbassert((scanPtr->m_agg_curr_batch_size_bytes % sizeof(Uint32)) == 0);
+  if (scanPtr->m_aggregation) {
+    /*
+     * Moz
+     * TODO (Zhao)
+     * potential crash here? double check.
+     * 1. API quits while scanning. scanPtr->scanState == WAIT_CLOSE_SCAN
+     * and scanPtr->m_agg_curr_batch_size_bytes maybe 0.
+     */
+    assert(scanPtr->m_agg_curr_batch_size_bytes ||
```

Comment:
assert -> ndbassert since in block code

Reply:
ndbrequire is assert that stays in production code, is fine to keep that if you like, but some of them might be ok to convert to ndbassert.

Reply:
ok